### DISCUSSION TITLE
Running validation & transformation on query results

### DISCUSSION THREAD

**Discussion by ian-schu** [prisma/prisma#22038] [Q&A] [answered]
### Question

In my codebase we have a `repo` layer where all Prisma queries live – we fetch results out of the DB and then do some validation / transformation of those records before handing them back to the application to treat as business objects. I'll give examples of such validation below.

I've understood recently [from this blog](https://www.prisma.io/blog/client-extensions-preview-8t3w27xkrxxn#example-transformed-fields) that we can do custom deserialization on any fields we like, including adding new computed fields. It looks like we can either transform existing fields while _retaining_ the same type, or we can add new fields of any type that we want.

This is helpful, but what I'd _really_ like to do is remap the whole result to a different type, where certain fields have been validated and transformed. **We do this today with plain old functions, my question is whether this can be done with a Prisma client extension instead.**

## Full example

Example schema:

```prisma
/// A simple / minimal example to demonstrate our validation / transformation needs
model PetAnimalRecord {
  id              BigInt    @id @default(autoincrement())
  name            String    @db.Text
  species         String    @db.Text
}
```

Here's how we would remap this today in our repo file:

```ts
// pet-animal-repo.ts

// First, the business object type
// This is a remapping of the raw Prisma type to an application-safe version
// The point is to transform Prisma fields to types that are more immediately usable by the app
export type PetAnimal = Omit<PetAnimal, 'id' | 'species'> & {
  id: PetAnimalId, // this type is a URL-safe wrapper around the raw database ID
  species: AnimalSpecies // string validated as an enum member
}

// Next, define a simple 'mapper' function - this takes the raw Prisma
// result and actually transforms it to the business object type we defined above
const toPetAnimal = (r: PetAnimalRecord): PetAnimal => ({
  ... r,
  id: PetAnimalId.parse(r.id),
  species: AnimalSpecies.parse(r.species) // suppose this is a zod parser
)}

// Now an example of a simple query function, utilizing the business object type
// and mapper function we defined above
export const getPetsByName = async (name: string): Promise<PetAnimal[]> => {
  const results = await prisma.petAnimalRecord.findMany({ where: { name }})
  return results.map(toPetAnimal)
}

```

## Remarks
This approach works fine enough and does not involve any magic. On the other hand, it's a fair amount of boilerplate (especially for larger models) and we have to apply `toPetAnimal` on every result that we get from Prisma. **The question is whether this validating / transforming can be absorbed into a Prisma client extension**.

**Comment by ludralph** [comment]
Hi @ian-schu 👋 

Thank you for raising this question.

> The question is whether this validating / transforming can be absorbed into a Prisma client extension.

Based on the example you shared, you should be able to perform the validation and transformation using the Prisma Client Extension.

You could do something like:
```js
const prisma = new PrismaClient().$extends({
  model: {
    petAnimalRecord: {
      async getPetsByName(name: string) {
        const pets = await prisma.petAnimalRecord.findMany({
          where: { name }
        })
        const petsAsPetAnimals = pets.map(toPetAnimal)
        return petsAsPetAnimals
      }
    }
  }
})
```
Then they can call it as follows:
```js
const result = await prisma.petAnimalRecord.getPetsByName("cat")
```
It should be possible to re-map the entire output types of the “default” Prisma Client queries as well.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by ian-schu** [reply]
Thanks @ludralph, I took a few minutes to look at this, and it appears that this 

Taking a closer look at this, I see two problems:
1. I would need to modify every operation independently, because if I do `$allOperations` then I'll be trying to run `toPetAnimal` on non-model results such as update counts.
2. Even when I copy / paste the remapping logic on built-in queries such as `findFirst`, at the call-site it does not appear that the TS compiler is aware of the changes.

Here's the short example that I tried:
```ts
const extendedClient = prisma.$extends({
  model: {
    petAnimalRecord: {
      $findUniqueOrThrow: async ({ args, query }) => {
        const result = await query(args);
        return toPetAnimal(result);
      },

      $findFirst: async ({ args, query }) => {
        const result = await query(args);
        return toPetAnimal(result);
      },
    },
  },
});
```

The way extensions work, the TS compiler doesn't seem to know that the return type for `findFirst` has been modified. This is evident from the fact that the `id` is still a `bigint`, the `species` field is still a `string`, etc.:


![image](https://github.com/prisma/prisma/assets/34523645/73e3a7f0-75df-41c4-b812-2d07c224ad6a)

**Reply by ian-schu** [reply]
@ludralph if this is still expected behavior ☝🏻 and it's not possible to dynamically change the return type of these built-in operations, then that's fine. I just want to determine if there is (or is not) a way to concisely handle this with Client Extensions, as of today.

**Comment by millsp** [comment]
Hey @ian-schu, what my colleague @ludralph has suggested you is also what I would usually prescribe. We don't really recommend overriding our own base functions, although you can and you might easily break things. So I did a little experiment today to see if with what we have today we could make it possible to map to any shape and make it typesafe.

I think I have found something that could match what you're looking for, take it with a pinch of salt, as I am slightly (but correctly) abusing Promises to achieve the desired result. So here we go, what do you think of this example? Does it work for you?

```ts
const prisma = new PrismaClient().$extends({
  result: {
    user: {
      then: { // this basically tells the JS runtime and TS that our result is a promise
              // this causes TS and JS to unfold the Promise until it gets its final result
        needs: { name: true, email: true },
        compute(data) {
          // here we implement the part of the Promise that interests us
          return (onfulfilled?: ((value: string /** Your type */) => any)) => {
            onfulfilled?.(`${data.name}${data.email}`)
          }
        }
      }
    }
  }
})

const email = `email@${Date.now()}`

// Now both TS and JS will resolve the promise, as our result has a `then` property
const user = await prisma.user.create({
    data: {
        email,
    },
})

console.log(user) // nullemail@1700598684048
```

Keep in mind that this is a workaround obviously, though it may work, I am also looking to learn from this. Feel free to open a feature request.

**Reply by millsp** [reply]
Additionally, a similar and less hacky approach would be to create a method on the result itself

```ts
const prisma = new PrismaClient().$extends({
  result: {
    user: {
      toSomethingElse: {
        needs: { name: true, email: true },
        compute(data) {
          return () => {
            return `${data.name}${data.email}`
          }
        }
      }
    }
  }
})

const email = `email@${Date.now()}`
const user = await prisma.user.create({
    data: {
        email,
    },
})

console.log(user.toSomethingElse()) // nullemail@1700598684048
```

**Reply by ian-schu** [reply]
> I think I have found something that could match what you're looking for, take it with a pinch of salt, as I am slightly (but correctly) abusing Promises to achieve the desired result.

@millsp very clever solution! I will play with this, though I'm not sure I want to to use a hacky solution to get this done. I guess I was hoping for more of a standard / recommended approach, but it sounds like my use case is not as common as I hoped.

Maybe you could answer a broader question for me: is it normal for developers to just use Prisma's own generated types as their "business objects" to get passed around the backend? I'm just curious about basic scenarios where you might want to validate something coming out of the database, such as the following:

1. You have a `@db.Text` column, but you want to validate that it contains a member of a Typescript enum (or `zod` enum).*
2. You have a `jsonb` column, and you want to validate that its contents conform to a `zod` object schema when you pull records out of the database.

These seem like very normal circumstances to me. Is it simply a matter of adding a computed value _alongside_ the raw Prisma field, and that's where you do the validation and type conversion? And then expecting that any callers will refer to the computed field instead of the raw value?

\* Yes I know that you can model enums directly in the database / Prisma schema, but I believe it's common for developers to choose _not_ to do this, and just leave the enum as a concern of the application.

**Reply by millsp** [reply, marked as answer]
>  I will play with this, though I'm not sure I want to to use a hacky solution to get this done

I totally get it, feel free to open a feature request or look if anything similar is open, I think it is worth it.

> Maybe you could answer a broader question for me: is it normal for developers to just use Prisma's own generated types as their "business objects" to get passed around the backend? I'm just curious about basic scenarios where you might want to validate something coming out of the database, such as the following:

If you need to do some extra validation on your data without extra steps like transforming the data to something completely different, you can use `query` extensions, if you have used our middleware feature before then its concept is very similar. I can definitely imagine a zod validator being implemented there. You can manage both input and output, you cannot change the type shape though. Would that work for you?

**Reply by ian-schu** [reply]
Thanks @millsp, I think you've covered everything I want to know in this area. I appreciate the help from you both!
